src/lsptoolshost/restore.ts

```diff
@@ -0,0 +1,108 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { RoslynLanguageServer } from './roslynLanguageServer';
+import { RestorableProjects, RestoreParams, RestorePartialResult, RestoreRequest } from './roslynProtocol';
+import path = require('path');
+
+let _restoreInProgress = false;
+
+export function registerRestoreCommands(
+    context: vscode.ExtensionContext,
+    languageServer: RoslynLanguageServer,
+    restoreChannel: vscode.OutputChannel
+) {
+    context.subscriptions.push(
+        vscode.commands.registerCommand('dotnet.restore.project', async (_request): Promise<void> => {
+            return chooseProjectToRestore(languageServer, restoreChannel);
+        })
+    );
+    context.subscriptions.push(
+        vscode.commands.registerCommand('dotnet.restore.all', async (): Promise<void> => {
+            return restore(languageServer, restoreChannel);
+        })
+    );
+}
+async function chooseProjectToRestore(
+    languageServer: RoslynLanguageServer,
+    restoreChannel: vscode.OutputChannel
+): Promise<void> {
+    const projects = await languageServer.sendRequest0(
+        RestorableProjects.type,
+        new vscode.CancellationTokenSource().token
```

Comment:
Is there not a better cancellation token to use for a command, or some UI that can be created for this? I don't think it's important to, as much as I am wondering out loud what the expected pattern is for things like this.

Reply:
We could put a progress bar on this and use cancellation there.  But it didn't seem necessary in this case - the request barely does anything and immediately shows a quick pick when its done.

Reply:
Yeah a progress bar does seem overkill. I wasn't sure if there was a simple pattern for commands with async work that's commonly used.

Reply:
I don't know of one unfortunately :(